Frameworks/CoreGraphics/DWriteFontBinaryDataCollectionLoader.mm

```diff
@@ -0,0 +1,186 @@
+
+    ComPtr<DWriteFontBinaryDataCollectionLoader> m_parent;
+};
+
+// DWriteFontBinaryDataCollectionLoader member functions
+
+DWriteFontBinaryDataCollectionLoader::DWriteFontBinaryDataCollectionLoader() {
+    m_fontDatasSet.reset(CFSetCreateMutable(nullptr, 0, &kCFTypeSetCallBacks));
+}
+
+HRESULT DWriteFontBinaryDataCollectionLoader::DWriteFontBinaryDataCollectionLoader::RuntimeClassInitialize() {
+    return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE DWriteFontBinaryDataCollectionLoader::CreateEnumeratorFromKey(_In_ IDWriteFactory* factory,
+                                                                                        _In_ const void* collectionKey,
+                                                                                        _In_ UINT32 collectionKeySize,
+                                                                                        _Out_ IDWriteFontFileEnumerator** enumerator) {
+    *enumerator = Make<DWriteFontFileEnumerator>(this).Detach();
+    return S_OK;
+}
+
+HRESULT DWriteFontBinaryDataCollectionLoader::AddDatas(CFArrayRef fontDatas, CFArrayRef* errors) {
+    // Don't bother creating this array if errors is null
+    CFMutableArrayRef outErrors = errors ? CFArrayCreateMutable(nullptr, 0, &kCFTypeArrayCallBacks) : nil;
+    HRESULT ret = S_OK;
+
+    CFIndex count = CFArrayGetCount(fontDatas);
+    for (CFIndex i = 0; i < count; ++i) {
+        CFDataRef data = static_cast<CFDataRef>(CFArrayGetValueAtIndex(fontDatas, i));
+        if (data) {
+            if (CFSetContainsValue(m_fontDatasSet.get(), data)) {
+                __AppendErrorIfExists(outErrors, kCTFontManagerErrorAlreadyRegistered);
+                ret = S_FALSE;
+            } else {
+                m_fontDatas.push_back({});
+
+                CFRetain(data);
+                m_fontDatas.back().first.reset(data);
+
+                CFSetAddValue(m_fontDatasSet.get(), data);
+                __AppendNullptrIfExists(outErrors);
+            }
+        } else {
+            __AppendErrorIfExists(outErrors, kCTFontManagerErrorInvalidFontData);
+            ret = S_FALSE;
+        }
+    }
+
+    if (errors) {
+        *errors = outErrors;
+    }
+
+    return ret;
+}
+
+HRESULT DWriteFontBinaryDataCollectionLoader::RemoveDatas(CFArrayRef fontDatas, CFArrayRef* errors) {
+    // Don't bother creating this array if errors is null
+    CFMutableArrayRef outErrors = errors ? CFArrayCreateMutable(nullptr, 0, &kCFTypeArrayCallBacks) : nil;
+    HRESULT ret = S_OK;
+
+    CFIndex count = CFArrayGetCount(fontDatas);
+    for (CFIndex i = 0; i < count; ++i) {
+        CFDataRef data = static_cast<CFDataRef>(CFArrayGetValueAtIndex(fontDatas, i));
+        if (data) {
+            if (CFSetContainsValue(m_fontDatasSet.get(), data)) {
+                CFSetRemoveValue(m_fontDatasSet.get(), data);
+
+                for (auto it = m_fontDatas.begin(); it != m_fontDatas.end(); ++it) {
+                    if (CFEqual(it->first.get(), data)) {
```

Comment:
Probably rather than the current setup (CFSet and vector<pair>), we should just use an std::unordered_map that uses CFData's hash/equal

---
In reply to: [91189304](https://github.com/Microsoft/WinObjC/pull/1504#discussion_r91189304) [](ancestors = 91189304,91176040)